firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +599,324 @@ def parenthesize(arg, prec=None, parent=None):
+                    callables_table)
+
+        def emit_call_insn(self, insn, target, expression_to_code_mapper):
+            assert self.is_ready_for_codegen()
+            assert isinstance(insn, loopy.CallInstruction)  # for batched this should be call instruction
+
+            parameters = insn.expression.parameters
+
+            if type(parameters) != list:
+                parameters = list(parameters)
+            par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]  # TODO: get the reads right
+
+            parameters.append(insn.assignees[0])
+            par_dtypes.append(self.arg_id_to_dtype[0])
+
+            mat_descr_A = self.arg_id_to_descr[0]
+
+            arg_c_parameters = [
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                if isinstance(par, SubArrayRef) else
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                for par, par_dtype in zip(parameters, par_dtypes)
+            ]
+            c_parameters = []
+            c_parameters.insert(0, arg_c_parameters[-1])  # out
+            c_parameters.insert(1, arg_c_parameters[0])  # A
+            c_parameters.insert(2, arg_c_parameters[1])  # B
+            c_parameters.insert(3, mat_descr_A.shape[1])  # n
+            return var(self.name_in_target)(*c_parameters), False
+
+        def generate_preambles(self, target):
+            assert isinstance(target, CTarget)
+            code = """#include <string.h>
+                #include <stdio.h>
+                #include <stdlib.h>
+
+                #ifndef Solve_HPP
+                #define Solve_HPP
+
+                void solve_(PetscScalar* out, PetscScalar* A, PetscScalar* B, PetscBLASInt N)
+                {
+                    PetscBLASInt info;
+                    PetscBLASInt* ipiv=(PetscBLASInt*) malloc(N*sizeof(PetscBLASInt));
+                    memcpy(out,B,N*sizeof(PetscScalar));
+                    PetscBLASInt NRHS;
+                    NRHS=1;
+                    LAPACKgesv_(&N,&NRHS,A,&N,ipiv,out,&N,&info);
+                    if(info!=0)
+                        fprintf(stderr,\"Gesv throws nonzero info.\");
+                }
+                #endif
+            """
+
+            yield("lapack_solve", "#include <petscsystypes.h>\n#include <petscblaslapack.h>\n"+code)
+            return
+
+    def fac_fn_lookup(target, identifier):
+        if identifier == 'solve':
+            return SolveCallable(name='solve')
+
+        return None
+
+    loopy_merged = loopy.register_function_id_to_in_knl_callable_mapper(loopy_merged, fac_fn_lookup)
+
+    return loopy_merged
+
+
 def slate_to_cpp(expr, temps, prec=None):
```

Comment:
This is now all still here because you still need "old" slate mode for some things, I guess.

Reply:
Yes I haven't adjusted the multigrid tests to be loopy compatible, yet.